commit ec9d3b52038c046a283c4b01eabeb3d58e696e99
Author: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date:   2019-09-18T18:19:12+02:00 (3 weeks ago)
    
    OS-7991 port_get() and port_getn() implementation disagrees with documentation

diff --git a/usr/src/head/port.h b/usr/src/head/port.h
index 8b09b39c8b..d784262669 100644
--- a/usr/src/head/port.h
+++ b/usr/src/head/port.h
@@ -45,8 +45,9 @@ int	port_associate(int, int, uintptr_t, int, void *);
 int	port_dissociate(int, int, uintptr_t);
 int	port_send(int, int, void *);
 int	port_sendn(int [], int [], uint_t, int, void *);
-int	port_get(int, port_event_t *, struct timespec *);
-int	port_getn(int, port_event_t [], uint_t, uint_t *, struct timespec *);
+int	port_get(int, port_event_t *, const struct timespec *);
+int	port_getn(int, port_event_t [], uint_t, uint_t *,
+    const struct timespec *);
 int	port_alert(int, int, int, void *);
 
 #ifdef	__cplusplus
diff --git a/usr/src/lib/libc/port/gen/event_port.c b/usr/src/lib/libc/port/gen/event_port.c
index df8f5c7a64..4dd50e4890 100644
--- a/usr/src/lib/libc/port/gen/event_port.c
+++ b/usr/src/lib/libc/port/gen/event_port.c
@@ -64,7 +64,7 @@ port_associate(int port, int source, uintptr_t object, int events, void *user)
 
 
 int
-port_get(int port, port_event_t *pe, struct timespec *to)
+port_get(int port, port_event_t *pe, const struct timespec *to)
 {
 	rval_t	r;
 	if (to)
@@ -77,7 +77,7 @@ port_get(int port, port_event_t *pe, struct timespec *to)
 
 int
 port_getn(int port, port_event_t list[], uint_t max, uint_t *nget,
-    struct timespec *timeout)
+    const struct timespec *timeout)
 {
 	rval_t	r;
 	if (nget == NULL) {
diff --git a/usr/src/uts/common/fs/portfs/port.c b/usr/src/uts/common/fs/portfs/port.c
index 04a2a421db..ebfd06ae22 100644
--- a/usr/src/uts/common/fs/portfs/port.c
+++ b/usr/src/uts/common/fs/portfs/port.c
@@ -615,6 +615,9 @@ portfs(int opcode, uintptr_t a0, uintptr_t a1, uintptr_t a2, uintptr_t a3,
 			error = port_getn(pp, (port_event_t *)a1, 1,
 			    (uint_t *)&nget, &port_timer);
 		} while (nget == 0 && error == 0 && port_timer.pgt_loop);
+
+		if (nget == 0 && error == 0 && a4 != 0)
+			error = ETIME;
 		break;
 	}
 	case	PORT_GETN:
